### ISSUE TITLE
csi-smb-controller: csi-provisioner (87) and csi-resizer (95) restart well above sibling baseline — likely leader-election cycling

### ISSUE THREAD

**Issue by jwilleke** [jwilleke/mj-infra-flux#43]
## Observation

`csi-smb-controller-6ff748dd57-r2kf2` (running since 2025-02-05) has four containers with very different restart counts:

| Container | Restarts | Note |
| --- | --- | --- |
| `csi-resizer` | **95** | leader-elected sidecar |
| `csi-provisioner` | **87** | leader-elected sidecar |
| `smb` | 59 | actual CSI driver, no leader election |
| `liveness-probe` | 55 | trivial sidecar, no leader election |

All four currently show `started=2026-05-04T08:24:1XZ` (post-host-reboot). They all see the same number of host reboots over the pod's lifetime (~10 in the relevant window per `last reboot`), so the host-reboot baseline is roughly the bottom row's count of 55–59. The differential — about **30–40 extra restarts on the leader-elected sidecars** — is the actual signal.

## What csi-smb does on this cluster

- StorageClass `smb` (provisioner `smb.csi.k8s.io`).
- One PV `pv-smb` (50 GiB, RWX, Retain), bound to `default/pvc-smb` for 341 days.
- Mounts `//192.168.68.24/docs` (`tower.nerdsbythehour.com`) using credentials in `default/smbcreds` (user `jim`).
- Consumer: `default/smb-filebrowser` — `Running 1/1 Ready`, 51 restarts in 169 d (matches the host-reboot baseline).

## Likely cause

`csi-provisioner` and `csi-resizer` are upstream Kubernetes [external-provisioner](https://github.com/kubernetes-csi/external-provisioner) and [external-resizer](https://github.com/kubernetes-csi/external-resizer) sidecars. By design, they **exit non-zero when leader-election fails** (lost lease, API server unreachable for the lease duration, etc.) — kubelet then restarts them. Each lease loss = +1 restart. The other two containers in the pod don't do leader election, so they stay up.

A tail of `csi-resizer` and `csi-provisioner` logs after the May 4 reboot shows clean startup with successful lease acquisition; the previous-instance log (`--previous`) shows the same shape from April 9. We can't see further back than two instances, so the bulk of the historical restarts are not directly visible — but the leader-election design plus the differential count is consistent with periodic API blips on a single-node k3s cluster (etcd compaction, brief network hiccups, k3s service restarts, kernel-update reboots).

## Why this might still be worth investigating

- ~30–40 extra restarts in ~5 months is roughly one leader-election cycle every 4 days. That's plausible noise on a single-node cluster, but a tighter cadence would suggest a real underlying issue (e.g., the SMB target `192.168.68.24` is currently unreachable — see side note — which doesn't affect the controller directly but is a cluster health hint).
- If a real provisioning or resize event happens to land in the brief gap between lease-loss and re-acquisition, the CSI request would be retried by kube-controller-manager but with extra latency.
- Memory limits are 400 Mi on each leader-elected sidecar; logs show no OOM activity. Resource pressure is not the cause.

## Acceptance / close criteria

This may turn out to be benign by-design behavior, in which case "won't fix" with a brief note is the right outcome. Investigation paths:

- [ ] Inspect `kube-system` Lease objects `external-resizer-smb-csi-k8s-io` and `smb-csi-k8s-io` history (or rebuild from etcd watch traces) to confirm lease churn rate.
- [ ] Watch the pod's restart counter for ~7 days and compute the new-restarts-per-day rate. If it's ≤ 1/day with no pod siblings restarting, this is normal CSI sidecar behavior — close as won't-fix with a comment.
- [ ] If the rate is higher, correlate with k3s service restarts (`systemctl status k3s -p NRestarts`), API server availability, and `journalctl -u k3s` errors. The actual fix would likely be on the k3s/API-availability side, not on csi-smb itself.

## Related

- The SMB target `tower.nerdsbythehour.com` (192.168.68.24) is currently unreachable (`ping` 100% loss, `nmblookup` no reply). This does not directly cause controller-sidecar restarts (those are API-only) but the smb-filebrowser pod is presumably holding a stale CIFS mount. Worth a separate bug if storage on that share is needed.
- Sibling cluster-wide flapping observation surfaced in session log 2026-05-05-11; cert-manager-cainjector at +28 over baseline is a similar shape and may warrant its own issue.

**Comment by jwilleke** [comment]
Closing as **obsolete** — the entire csi-smb stack has been removed from the cluster. The leader-election cycling that filed this issue is moot because there is no driver left to cycle.

## What happened

The investigation surfaced a much bigger finding than the original 30–40 extra restarts:

- The csi-smb driver was provisioning storage for `default/pvc-smb` (50 GiB RWX, bound to `pv-smb`, pointed at `//192.168.68.24/docs` = `tower.nerdsbythehour.com`).
- **`tower.nerdsbythehour.com` no longer exists** — the host is gone.
- **`pvc-smb` had zero consumers** cluster-wide. Nothing mounted it.
- The `default/smb-filebrowser` deployment, despite the name, used `emptyDir` for both `/config` and `/srv` — it was never actually wired to the SMB share. State reset on every restart; no traffic for the past 24h+ (`filebrowser.nerdsbythehour.com` had 0 hits in 5000 traefik log lines).
- Net: the driver was provisioning storage for a PVC nobody used, from a host that doesn't exist, to support a deployment whose volume mounts didn't include that storage anyway.

## Cleanup landed today

- **Phase A** — `mj-infra-flux@f60a8a5` removed `apps/production/filebrowser/` (Deployment + Service + Ingress) and pruned the related references from `apps/production/authentik/README.md`. Flux reconciled and pruned the workload.
- **Phase B** — direct `kubectl delete` for the four hand-applied resources: `pvc-smb`, `pv-smb`, `StorageClass smb`, `secret/smbcreds`.
- **Phase C** — direct `kubectl delete` for the driver itself: 13 resources (Deployment + 2 DaemonSets + CSIDriver + 3 ClusterRoles + 3 ClusterRoleBindings + 2 ServiceAccounts + 2 leader-election Leases).

## Verification

```text
kubectl get csidriver smb.csi.k8s.io               → NotFound
kubectl get deploy -n kube-system csi-smb-controller → NotFound
kubectl get ds -n kube-system csi-smb-node          → NotFound
kubectl get clusterrole,clusterrolebinding | grep smb  → empty
kubectl get all -n kube-system | grep smb              → empty
```

## Side note for future reference

`csi-provisioner` and `csi-resizer` exiting on lease loss is **by design**, not a bug — the upstream sidecars treat lease loss as a safety signal and rely on kubelet to restart them. So if you ever bring SMB storage back (or any other CSI driver), don't be alarmed by sidecars accumulating ~1 restart per 4 days from leader-election re-elections; that's normal behavior on a single-node k3s cluster where API blips happen periodically (etcd compaction, brief networking events, k3s service restarts).
